src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java

```diff
@@ -1,155 +1,162 @@
+     * every missing product, category, productType or custom object Id in a combined request. For
+     * each fetched key/id pair, the method will insert it into the {@code idToKey} cache and then
+     * return the cache in a {@link CompletableFuture} after the request is successful.
+     *
+     * @param productIds      the product ids to find a key mapping for.
+     * @param categoryIds     the category ids to find a key mapping for.
+     * @param productTypeIds  the productType ids to find a key mapping for.
+     * @param customObjectIds the custom object ids to find a key mapping for.
+     * @return a map of id to key representing products, categories, productTypes and customObjects in
+     * the CTP project defined by the injected {@code ctpClient}.
+     */
+    @Nonnull
+    @Override
+    public CompletionStage<Map<String, String>> getIdToKeys(
+        @Nonnull final Set<String> productIds,
+        @Nonnull final Set<String> categoryIds,
+        @Nonnull final Set<String> productTypeIds,
+        @Nonnull final Set<String> customObjectIds) {
+
+        final Set<String> nonCachedProductIds = getNonCachedIds(productIds);
+        final Set<String> nonCachedCategoryIds = getNonCachedIds(categoryIds);
+        final Set<String> nonCachedProductTypeIds = getNonCachedIds(productTypeIds);
+        final Set<String> nonCachedCustomObjectIds = getNonCachedIds(customObjectIds);
+
+        if (nonCachedProductIds.isEmpty()
+            && nonCachedCategoryIds.isEmpty()
+            && nonCachedProductTypeIds.isEmpty()) {
+            return fetchCustomObjectKeys(customObjectIds);
+        }
+
+        int CHUNK_SIZE = 500;
+        List<List<String>> productIdsChunk = ChunkUtils.chunk(nonCachedProductIds, CHUNK_SIZE);
+        List<List<String>> categoryIdsChunk = ChunkUtils.chunk(nonCachedCategoryIds, CHUNK_SIZE);
+        List<List<String>> productTypeIdsChunk = ChunkUtils.chunk(nonCachedProductTypeIds, CHUNK_SIZE);
+
+        List<ResourceIdsGraphQlRequest> collectedRequests = new ArrayList<>();
+
+        collectedRequests.addAll(
+            createResourceIdsGraphQlRequests(productIdsChunk, GraphQlQueryResources.PRODUCTS));
+        collectedRequests.addAll(
+            createResourceIdsGraphQlRequests(categoryIdsChunk, GraphQlQueryResources.CATEGORIES));
+        collectedRequests.addAll(
+            createResourceIdsGraphQlRequests(productTypeIdsChunk, GraphQlQueryResources.PRODUCT_TYPES));
+
+        // TODO: Adapt to run grapqhl and rest call in parallel.
+        return ChunkUtils.executeChunks(getCtpClient(), collectedRequests)
+                         .thenApply(ChunkUtils::flattenGraphQLBaseResults)
+                         .thenApply(results -> {
+                             cacheKeys(results);
+                             return allResourcesIdToKey;
+                         })
+                         .thenCompose(ignored -> fetchCustomObjectKeys(nonCachedCustomObjectIds));
```

Comment:
here we need to chunk custom object query too, right?

Reply:
I already implemented similar on UnresolvedReferencesServiceImpl.fetch, you might inspire:

https://github.com/commercetools/commercetools-sync-java/blob/a949fc1a2b8ce24c160d7d211234fd5bfab4ea7b/src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java#L56-L87

Reply:
fb7b913